homeassistant/components/xiaomi_miio/fan.py

```diff
@@ -386,15 +317,48 @@ async def async_turn_off(self, **kwargs) -> None:
             self.async_write_ha_state()
 
 
-class XiaomiAirPurifier(XiaomiGenericDevice):
-    """Representation of a Xiaomi Air Purifier."""
+class XiaomiGenericAirPurifier(XiaomiGenericDevice):
+    """Representation of a generic AirPurifier device."""
 
-    PRESET_MODE_MAPPING = {
-        "Auto": AirpurifierOperationMode.Auto,
-        "Silent": AirpurifierOperationMode.Silent,
-        "Favorite": AirpurifierOperationMode.Favorite,
-        "Idle": AirpurifierOperationMode.Favorite,
-    }
+    def __init__(self, name, device, entry, unique_id, coordinator):
+        """Initialize the generic AirPurifier device."""
+        super().__init__(name, device, entry, unique_id, coordinator)
+
+        self._speed_count = 100
+
+    @property
+    @abstractmethod
+    def operation_mode_class(self):
+        """Hold operation mode class."""
+
+    @property
+    def speed_count(self):
+        """Return the number of speeds of the fan supported."""
+        return self._speed_count
+
+    @property
+    def preset_mode(self):
+        """Get the active preset mode."""
+        if self._state:
+            preset_mode = self.operation_mode_class(self._mode).name
+            return preset_mode if preset_mode in self._preset_modes else None
+
+        return None
+
+    @callback
+    def _handle_coordinator_update(self):
+        """Fetch state from the device."""
+        self._state = self.coordinator.data.is_on
+        self._state_attrs.update(
```

Comment:
Where are the state attributes updated to something else than `None`?

Reply:
What a stupid mistake 🤦‍♂️